File: storage/internal/disk/superblock.go
// Package disk — superblock.
//
// Architecture note (A4-Metadata-As-Chunks, docs/02 "Bootstrap", docs/14 S14):
//
// The superblock is the ONLY non-chunk data on any NovaNas disk. It solves
// the chicken-and-egg bootstrap problem: the metadata service lives on
// chunks, but to find those chunks the cluster must first enumerate its
// disks and learn which pool they belong to. Each disk therefore carries a
// small (4 KiB), fixed-offset, CRC-protected descriptor at the start of the
// device.
//
// Layout (little-endian, fixed offsets):
//
//	offset  size  field
//	------  ----  -------------------------------------------------
//	0       8     magic   ("NOVANAS\x00")
//	8       4     version (currently 1)
//	12      4     flags   (reserved, 0)
//	16      16    disk UUID (binary, matches WWN-derived identity)
//	32      32    pool ID  (utf-8, zero-padded)
//	64      4     role     (1 = data, 2 = metadata, 3 = both)
//	68      32    CRUSH map digest (sha256 of canonical CRUSH map)
//	100     32    metadata volume name (utf-8, zero-padded)
//	132     64    metadata volume root chunk ID (hex sha256, zero-padded)
//	196     8     metadata volume version (monotonic counter)
//	204     8     created_unix_nanos
//	212     8     updated_unix_nanos
//	220     3872  reserved (zeroed)
//	4092    4     crc32c over bytes [0..4092)
//
// The superblock is written at byte offset 0 of the raw device. For GPT
// co-existence, higher deployments may opt to shift the superblock to a
// reserved area; the WriteAt/ReadAt offset is centralised in
// SuperblockOffset for that reason.
package disk

import (
	"encoding/binary"
	"errors"
	"fmt"
	"hash/crc32"
	"os"
	"time"
)

// SuperblockSize is the on-disk size of the superblock (4 KiB).
const SuperblockSize = 4096

// SuperblockOffset is the byte offset on the raw device where the
// superblock lives. Zero by default; shift if you need to leave room for
// a partition table.
const SuperblockOffset int64 = 0

// SuperblockMagic is the magic number ("NOVANAS\x00"). 8 bytes.
var SuperblockMagic = [8]byte{'N', 'O', 'V', 'A', 'N', 'A', 'S', 0}

// SuperblockVersion is the current on-disk format version.
const SuperblockVersion uint32 = 1

// DiskRole is the role a disk plays in its pool.
type DiskRole uint32

const (
	// DiskRoleUnknown is the zero value; invalid at runtime.
	DiskRoleUnknown DiskRole = 0
	// DiskRoleData participates only in data pools.
	DiskRoleData DiskRole = 1
	// DiskRoleMetadata participates only in the metadata pool.
	DiskRoleMetadata DiskRole = 2
	// DiskRoleBoth participates in both data and metadata pools.
	DiskRoleBoth DiskRole = 3
)

// Superblock is the in-memory form of the on-disk superblock.
type Superblock struct {
	Version  uint32
	Flags    uint32
	DiskUUID [16]byte
	PoolID   string // max 32 bytes when serialised
	Role     DiskRole

	// CRUSHDigest is a SHA-256 (32 bytes) of the canonical CRUSH map. Used
	// to detect divergence across disks when a meta-quorum is assembled.
	CRUSHDigest [32]byte

	// Metadata-volume locator: how to reconstruct the metadata BlockVolume.
	MetaVolumeName      string // max 32 bytes
	MetaVolumeRootChunk string // max 64 bytes (hex sha256)
	MetaVolumeVersion   uint64

	CreatedUnixNanos int64
	UpdatedUnixNanos int64
}

// Field offsets — keep in sync with the layout comment above.
const (
	sbOffMagic          = 0
	sbOffVersion        = 8
	sbOffFlags          = 12
	sbOffDiskUUID       = 16
	sbOffPoolID         = 32
	sbOffRole           = 64
	sbOffCRUSHDigest    = 68
	sbOffMetaVolName    = 100
	sbOffMetaVolRoot    = 132
	sbOffMetaVolVersion = 196
	sbOffCreated        = 204
	sbOffUpdated        = 212
	sbOffCRC            = 4092

	sbFieldPoolIDLen      = 32
	sbFieldMetaVolNameLen = 32
	sbFieldMetaVolRootLen = 64
)

// Errors.
var (
	ErrSuperblockBadMagic   = errors.New("superblock magic mismatch")
	ErrSuperblockBadCRC     = errors.New("superblock CRC mismatch")
	ErrSuperblockBadVersion = errors.New("superblock version unsupported")
	ErrFieldTooLong         = errors.New("field exceeds its on-disk capacity")
)

// Marshal returns the fixed-size 4096-byte on-disk representation.
func (s *Superblock) Marshal() ([]byte, error) {
	if len(s.PoolID) > sbFieldPoolIDLen {
		return nil, fmt.Errorf("%w: pool id (%d > %d)", ErrFieldTooLong, len(s.PoolID), sbFieldPoolIDLen)
	}
	if len(s.MetaVolumeName) > sbFieldMetaVolNameLen {
		return nil, fmt.Errorf("%w: meta volume name", ErrFieldTooLong)
	}
	if len(s.MetaVolumeRootChunk) > sbFieldMetaVolRootLen {
		return nil, fmt.Errorf("%w: meta volume root chunk id", ErrFieldTooLong)
	}

	buf := make([]byte, SuperblockSize)
	copy(buf[sbOffMagic:sbOffMagic+8], SuperblockMagic[:])
	binary.LittleEndian.PutUint32(buf[sbOffVersion:], s.Version)
	binary.LittleEndian.PutUint32(buf[sbOffFlags:], s.Flags)
	copy(buf[sbOffDiskUUID:sbOffDiskUUID+16], s.DiskUUID[:])
	copy(buf[sbOffPoolID:sbOffPoolID+sbFieldPoolIDLen], []byte(s.PoolID))
	binary.LittleEndian.PutUint32(buf[sbOffRole:], uint32(s.Role))
	copy(buf[sbOffCRUSHDigest:sbOffCRUSHDigest+32], s.CRUSHDigest[:])
	copy(buf[sbOffMetaVolName:sbOffMetaVolName+sbFieldMetaVolNameLen], []byte(s.MetaVolumeName))
	copy(buf[sbOffMetaVolRoot:sbOffMetaVolRoot+sbFieldMetaVolRootLen], []byte(s.MetaVolumeRootChunk))
	binary.LittleEndian.PutUint64(buf[sbOffMetaVolVersion:], s.MetaVolumeVersion)
	binary.LittleEndian.PutUint64(buf[sbOffCreated:], uint64(s.CreatedUnixNanos))
	binary.LittleEndian.PutUint64(buf[sbOffUpdated:], uint64(s.UpdatedUnixNanos))

	table := crc32.MakeTable(crc32.Castagnoli)
	crc := crc32.Checksum(buf[:sbOffCRC], table)
	binary.LittleEndian.PutUint32(buf[sbOffCRC:], crc)
	return buf, nil
}

// UnmarshalSuperblock parses a 4096-byte buffer into a Superblock.
// Returns ErrSuperblockBadMagic / ErrSuperblockBadCRC / ErrSuperblockBadVersion
// on validation failure.
func UnmarshalSuperblock(buf []byte) (*Superblock, error) {
	if len(buf) != SuperblockSize {
		return nil, fmt.Errorf("superblock buffer size %d, want %d", len(buf), SuperblockSize)
	}
	var magic [8]byte
	copy(magic[:], buf[sbOffMagic:sbOffMagic+8])
	if magic != SuperblockMagic {
		return nil, ErrSuperblockBadMagic
	}

	table := crc32.MakeTable(crc32.Castagnoli)
	gotCRC := binary.LittleEndian.Uint32(buf[sbOffCRC:])
	wantCRC := crc32.Checksum(buf[:sbOffCRC], table)
	if gotCRC != wantCRC {
		return nil, fmt.Errorf("%w: got=%08x want=%08x", ErrSuperblockBadCRC, gotCRC, wantCRC)
	}

	s := &Superblock{
		Version: binary.LittleEndian.Uint32(buf[sbOffVersion:]),
		Flags:   binary.LittleEndian.Uint32(buf[sbOffFlags:]),
	}
	if s.Version != SuperblockVersion {
		return nil, fmt.Errorf("%w: got=%d want=%d", ErrSuperblockBadVersion, s.Version, SuperblockVersion)
	}
	copy(s.DiskUUID[:], buf[sbOffDiskUUID:sbOffDiskUUID+16])
	s.PoolID = trimZero(buf[sbOffPoolID : sbOffPoolID+sbFieldPoolIDLen])
	s.Role = DiskRole(binary.LittleEndian.Uint32(buf[sbOffRole:]))
	copy(s.CRUSHDigest[:], buf[sbOffCRUSHDigest:sbOffCRUSHDigest+32])
	s.MetaVolumeName = trimZero(buf[sbOffMetaVolName : sbOffMetaVolName+sbFieldMetaVolNameLen])
	s.MetaVolumeRootChunk = trimZero(buf[sbOffMetaVolRoot : sbOffMetaVolRoot+sbFieldMetaVolRootLen])
	s.MetaVolumeVersion = binary.LittleEndian.Uint64(buf[sbOffMetaVolVersion:])
	s.CreatedUnixNanos = int64(binary.LittleEndian.Uint64(buf[sbOffCreated:]))
	s.UpdatedUnixNanos = int64(binary.LittleEndian.Uint64(buf[sbOffUpdated:]))
	return s, nil
}

// trimZero strips trailing NUL bytes from a fixed-width utf-8 field.
func trimZero(b []byte) string {
	for i, c := range b {
		if c == 0 {
			return string(b[:i])
		}
	}
	return string(b)
}

// WriteSuperblock writes the superblock to the device at SuperblockOffset.
// Fills in Version / UpdatedUnixNanos automatically. Opens the device with
// O_WRONLY; caller is responsible for not writing to a mounted filesystem
// (use a dedicated raw disk).
func WriteSuperblock(devicePath string, s *Superblock) error {
	if s.Version == 0 {
		s.Version = SuperblockVersion
	}
	now := time.Now().UnixNano()
	if s.CreatedUnixNanos == 0 {
		s.CreatedUnixNanos = now
	}
	s.UpdatedUnixNanos = now

	buf, err := s.Marshal()
	if err != nil {
		return err
	}
	f, err := os.OpenFile(devicePath, os.O_WRONLY, 0)
	if err != nil {
		return fmt.Errorf("opening %s for superblock write: %w", devicePath, err)
	}
	defer f.Close()
	if _, err := f.WriteAt(buf, SuperblockOffset); err != nil {
		return fmt.Errorf("writing superblock to %s: %w", devicePath, err)
	}
	if err := f.Sync(); err != nil {
		return fmt.Errorf("syncing superblock on %s: %w", devicePath, err)
	}
	return nil
}

// ReadSuperblock reads and validates the superblock at SuperblockOffset.
func ReadSuperblock(devicePath string) (*Superblock, error) {
	f, err := os.Open(devicePath)
	if err != nil {
		return nil, fmt.Errorf("opening %s for superblock read: %w", devicePath, err)
	}
	defer f.Close()
	buf := make([]byte, SuperblockSize)
	if _, err := f.ReadAt(buf, SuperblockOffset); err != nil {
		return nil, fmt.Errorf("reading superblock from %s: %w", devicePath, err)
	}
	return UnmarshalSuperblock(buf)
}

File: storage/dataplane/src/backend/superblock.rs
//! Disk superblock (A4-Metadata-As-Chunks, docs/02 "Bootstrap").
//!
//! The superblock is the only non-chunk data on a NovaNas disk. It
//! carries the minimum state required to bootstrap the metadata
//! service without access to any prior metadata: disk identity, pool
//! membership, CRUSH-map digest, and a locator for the metadata
//! BlockVolume's root chunk.
//!
//! Byte layout: identical to the Go implementation in
//! `storage/internal/disk/superblock.go`; see that file for the
//! canonical field-offset table. Endianness: little-endian. Size: 4096
//! bytes. A CRC32C checksum over the first 4092 bytes lives at
//! offset 4092.
//!
//! This Rust module is intentionally pure (no SPDK) so it builds in
//! CI without the data-plane feature set. The raw-disk backend will
//! call `read_superblock` / `write_superblock` against an SPDK bdev in
//! production.

use crc32c::crc32c;
use std::fs::OpenOptions;
use std::io::{Read, Seek, SeekFrom, Write};
use std::path::Path;

/// On-disk superblock size (4 KiB).
pub const SUPERBLOCK_SIZE: usize = 4096;

/// Byte offset of the superblock within a raw device.
pub const SUPERBLOCK_OFFSET: u64 = 0;

/// Current superblock format version.
pub const SUPERBLOCK_VERSION: u32 = 1;

/// Magic bytes identifying a NovaNas superblock.
pub const SUPERBLOCK_MAGIC: [u8; 8] = *b"NOVANAS\0";

// Field offsets — keep in sync with the Go implementation.
const OFF_MAGIC: usize = 0;
const OFF_VERSION: usize = 8;
const OFF_FLAGS: usize = 12;
const OFF_DISK_UUID: usize = 16;
const OFF_POOL_ID: usize = 32;
const OFF_ROLE: usize = 64;
const OFF_CRUSH_DIGEST: usize = 68;
const OFF_META_VOL_NAME: usize = 100;
const OFF_META_VOL_ROOT: usize = 132;
const OFF_META_VOL_VERSION: usize = 196;
const OFF_CREATED: usize = 204;
const OFF_UPDATED: usize = 212;
const OFF_CRC: usize = 4092;

const POOL_ID_LEN: usize = 32;
const META_VOL_NAME_LEN: usize = 32;
const META_VOL_ROOT_LEN: usize = 64;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum DiskRole {
    Unknown = 0,
    Data = 1,
    Metadata = 2,
    Both = 3,
}

impl DiskRole {
    pub fn from_u32(v: u32) -> Self {
        match v {
            1 => DiskRole::Data,
            2 => DiskRole::Metadata,
            3 => DiskRole::Both,
            _ => DiskRole::Unknown,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Superblock {
    pub version: u32,
    pub flags: u32,
    pub disk_uuid: [u8; 16],
    pub pool_id: String,
    pub role: DiskRole,
    pub crush_digest: [u8; 32],
    pub meta_volume_name: String,
    pub meta_volume_root_chunk: String,
    pub meta_volume_version: u64,
    pub created_unix_nanos: i64,
    pub updated_unix_nanos: i64,
}

impl Default for Superblock {
    fn default() -> Self {
        Self {
            version: SUPERBLOCK_VERSION,
            flags: 0,
            disk_uuid: [0; 16],
            pool_id: String::new(),
            role: DiskRole::Unknown,
            crush_digest: [0; 32],
            meta_volume_name: String::new(),
            meta_volume_root_chunk: String::new(),
            meta_volume_version: 0,
            created_unix_nanos: 0,
            updated_unix_nanos: 0,
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum SuperblockError {
    #[error("superblock magic mismatch")]
    BadMagic,
    #[error("superblock CRC mismatch: got {got:08x} want {want:08x}")]
    BadCrc { got: u32, want: u32 },
    #[error("unsupported superblock version {0}")]
    BadVersion(u32),
    #[error("field {name} exceeds on-disk capacity ({len} > {max})")]
    FieldTooLong {
        name: &'static str,
        len: usize,
        max: usize,
    },
    #[error("buffer size {got} does not match SUPERBLOCK_SIZE {want}")]
    BadSize { got: usize, want: usize },
    #[error("io: {0}")]
    Io(#[from] std::io::Error),
}

fn put_fixed_str(
    buf: &mut [u8],
    offset: usize,
    len: usize,
    s: &str,
    name: &'static str,
) -> Result<(), SuperblockError> {
    if s.len() > len {
        return Err(SuperblockError::FieldTooLong {
            name,
            len: s.len(),
            max: len,
        });
    }
    let slot = &mut buf[offset..offset + len];
    for b in slot.iter_mut() {
        *b = 0;
    }
    slot[..s.len()].copy_from_slice(s.as_bytes());
    Ok(())
}

fn read_fixed_str(buf: &[u8], offset: usize, len: usize) -> String {
    let slot = &buf[offset..offset + len];
    let end = slot.iter().position(|&b| b == 0).unwrap_or(len);
    String::from_utf8_lossy(&slot[..end]).into_owned()
}

impl Superblock {
    /// Serialise to a 4096-byte buffer with CRC populated.
    pub fn marshal(&self) -> Result<[u8; SUPERBLOCK_SIZE], SuperblockError> {
        let mut buf = [0u8; SUPERBLOCK_SIZE];
        buf[OFF_MAGIC..OFF_MAGIC + 8].copy_from_slice(&SUPERBLOCK_MAGIC);
        buf[OFF_VERSION..OFF_VERSION + 4].copy_from_slice(&self.version.to_le_bytes());
        buf[OFF_FLAGS..OFF_FLAGS + 4].copy_from_slice(&self.flags.to_le_bytes());
        buf[OFF_DISK_UUID..OFF_DISK_UUID + 16].copy_from_slice(&self.disk_uuid);
        put_fixed_str(&mut buf, OFF_POOL_ID, POOL_ID_LEN, &self.pool_id, "pool_id")?;
        buf[OFF_ROLE..OFF_ROLE + 4].copy_from_slice(&(self.role as u32).to_le_bytes());
        buf[OFF_CRUSH_DIGEST..OFF_CRUSH_DIGEST + 32].copy_from_slice(&self.crush_digest);
        put_fixed_str(
            &mut buf,
            OFF_META_VOL_NAME,
            META_VOL_NAME_LEN,
            &self.meta_volume_name,
            "meta_volume_name",
        )?;
        put_fixed_str(
            &mut buf,
            OFF_META_VOL_ROOT,
            META_VOL_ROOT_LEN,
            &self.meta_volume_root_chunk,
            "meta_volume_root_chunk",
        )?;
        buf[OFF_META_VOL_VERSION..OFF_META_VOL_VERSION + 8]
            .copy_from_slice(&self.meta_volume_version.to_le_bytes());
        buf[OFF_CREATED..OFF_CREATED + 8].copy_from_slice(&self.created_unix_nanos.to_le_bytes());
        buf[OFF_UPDATED..OFF_UPDATED + 8].copy_from_slice(&self.updated_unix_nanos.to_le_bytes());

        let crc = crc32c(&buf[..OFF_CRC]);
        buf[OFF_CRC..OFF_CRC + 4].copy_from_slice(&crc.to_le_bytes());
        Ok(buf)
    }

    pub fn unmarshal(buf: &[u8]) -> Result<Self, SuperblockError> {
        if buf.len() != SUPERBLOCK_SIZE {
            return Err(SuperblockError::BadSize {
                got: buf.len(),
                want: SUPERBLOCK_SIZE,
            });
        }
        if buf[OFF_MAGIC..OFF_MAGIC + 8] != SUPERBLOCK_MAGIC {
            return Err(SuperblockError::BadMagic);
        }
        let got_crc = u32::from_le_bytes(buf[OFF_CRC..OFF_CRC + 4].try_into().unwrap());
        let want_crc = crc32c(&buf[..OFF_CRC]);
        if got_crc != want_crc {
            return Err(SuperblockError::BadCrc {
                got: got_crc,
                want: want_crc,
            });
        }
        let version = u32::from_le_bytes(buf[OFF_VERSION..OFF_VERSION + 4].try_into().unwrap());
        if version != SUPERBLOCK_VERSION {
            return Err(SuperblockError::BadVersion(version));
        }
        let flags = u32::from_le_bytes(buf[OFF_FLAGS..OFF_FLAGS + 4].try_into().unwrap());
        let mut disk_uuid = [0u8; 16];
        disk_uuid.copy_from_slice(&buf[OFF_DISK_UUID..OFF_DISK_UUID + 16]);
        let pool_id = read_fixed_str(buf, OFF_POOL_ID, POOL_ID_LEN);
        let role = DiskRole::from_u32(u32::from_le_bytes(
            buf[OFF_ROLE..OFF_ROLE + 4].try_into().unwrap(),
        ));
        let mut crush_digest = [0u8; 32];
        crush_digest.copy_from_slice(&buf[OFF_CRUSH_DIGEST..OFF_CRUSH_DIGEST + 32]);
        let meta_volume_name = read_fixed_str(buf, OFF_META_VOL_NAME, META_VOL_NAME_LEN);
        let meta_volume_root_chunk = read_fixed_str(buf, OFF_META_VOL_ROOT, META_VOL_ROOT_LEN);
        let meta_volume_version = u64::from_le_bytes(
            buf[OFF_META_VOL_VERSION..OFF_META_VOL_VERSION + 8]
                .try_into()
                .unwrap(),
        );
        let created_unix_nanos =
            i64::from_le_bytes(buf[OFF_CREATED..OFF_CREATED + 8].try_into().unwrap());
        let updated_unix_nanos =
            i64::from_le_bytes(buf[OFF_UPDATED..OFF_UPDATED + 8].try_into().unwrap());
        Ok(Self {
            version,
            flags,
            disk_uuid,
            pool_id,
            role,
            crush_digest,
            meta_volume_name,
            meta_volume_root_chunk,
            meta_volume_version,
            created_unix_nanos,
            updated_unix_nanos,
        })
    }
}

/// Write the superblock at byte offset SUPERBLOCK_OFFSET on `device`.
pub fn write_superblock(device: impl AsRef<Path>, sb: &Superblock) -> Result<(), SuperblockError> {
    let buf = sb.marshal()?;
    let mut f = OpenOptions::new().write(true).open(device.as_ref())?;
    f.seek(SeekFrom::Start(SUPERBLOCK_OFFSET))?;
    f.write_all(&buf)?;
    f.sync_all()?;
    Ok(())
}

/// Read and validate the superblock at SUPERBLOCK_OFFSET on `device`.
pub fn read_superblock(device: impl AsRef<Path>) -> Result<Superblock, SuperblockError> {
    let mut f = OpenOptions::new().read(true).open(device.as_ref())?;
    f.seek(SeekFrom::Start(SUPERBLOCK_OFFSET))?;
    let mut buf = [0u8; SUPERBLOCK_SIZE];
    f.read_exact(&mut buf)?;
    Superblock::unmarshal(&buf)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    use std::io::Write;

    fn sample() -> Superblock {
        let mut sb = Superblock {
            pool_id: "pool-default".into(),
            role: DiskRole::Both,
            meta_volume_name: "meta-vol".into(),
            meta_volume_root_chunk: "abc123".into(),
            meta_volume_version: 42,
            ..Superblock::default()
        };
        for i in 0..16 {
            sb.disk_uuid[i] = (i + 1) as u8;
        }
        for i in 0..32 {
            sb.crush_digest[i] = (i * 3) as u8;
        }
        sb
    }

    #[test]
    fn marshal_roundtrip() {
        let sb = sample();
        let buf = sb.marshal().unwrap();
        assert_eq!(buf.len(), SUPERBLOCK_SIZE);
        let got = Superblock::unmarshal(&buf).unwrap();
        assert_eq!(got.pool_id, sb.pool_id);
        assert_eq!(got.role, sb.role);
        assert_eq!(got.meta_volume_name, sb.meta_volume_name);
        assert_eq!(got.meta_volume_root_chunk, sb.meta_volume_root_chunk);
        assert_eq!(got.meta_volume_version, sb.meta_volume_version);
        assert_eq!(got.disk_uuid, sb.disk_uuid);
        assert_eq!(got.crush_digest, sb.crush_digest);
    }

    #[test]
    fn bad_magic() {
        let sb = sample();
        let mut buf = sb.marshal().unwrap();
        buf[0] ^= 0xff;
        assert!(matches!(
            Superblock::unmarshal(&buf),
            Err(SuperblockError::BadMagic)
        ));
    }

    #[test]
    fn bad_crc() {
        let sb = sample();
        let mut buf = sb.marshal().unwrap();
        buf[OFF_POOL_ID] ^= 0x01;
        assert!(matches!(
            Superblock::unmarshal(&buf),
            Err(SuperblockError::BadCrc { .. })
        ));
    }

    #[test]
    fn field_too_long() {
        let mut sb = sample();
        sb.pool_id = "x".repeat(POOL_ID_LEN + 1);
        assert!(matches!(
            sb.marshal(),
            Err(SuperblockError::FieldTooLong { .. })
        ));
    }

    #[test]
    fn write_read_file() {
        let dir = tempfile::tempdir().unwrap();
        let path = dir.path().join("fake-disk");
        {
            let mut f = File::create(&path).unwrap();
            f.write_all(&vec![0u8; 64 * 1024]).unwrap();
        }
        let sb = sample();
        write_superblock(&path, &sb).unwrap();
        let got = read_superblock(&path).unwrap();
        assert_eq!(got.pool_id, sb.pool_id);
        assert_eq!(got.meta_volume_root_chunk, sb.meta_volume_root_chunk);
    }
}
